{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Tutorial - Part 3\n",
    "\n",
    "---\n",
    "\n",
    "# Try/Except\n",
    "\n",
    "---\n",
    "\n",
    "Python provides a method to trap errors with the code at run time to prevent the script crashing. For example, the following program will prompt for a number and  print it out. It will continue prompting until the exit condition is reached.\n",
    "\n",
    "It takes a string as input and converts it to an integer.\n",
    "\n",
    "There is a problem with this code in that if text other than a number was entered, for example \"2d\", the script would crash. Try running the code with numbers and non-numeric entries:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type Control C or -1 to exit\n",
      "Enter a number: 1\n",
      "You entered: 1\n",
      "You entered an odd number\n",
      "Enter a number: 2\n",
      "You entered: 2\n",
      "You entered an even number\n",
      "Enter a number: -1\n",
      "You entered: -1\n",
      "You entered an odd number\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print (\"Type Control C or -1 to exit\")\n",
    "number = 1\n",
    "while number != -1:\n",
    "    number = int(input(\"Enter a number: \"))\n",
    "    print (\"You entered:\", number)\n",
    "    if number % 2 == 0:\n",
    "        print (\"You entered an even number\")\n",
    "    else:\n",
    "        print (\"You entered an odd number\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "\n",
    "This can be prevented using try/except:\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type Control C or -1 to exit\n",
      "Enter a number: g\n",
      "That was not a number\n",
      "Enter a number: 5\n",
      "You entered: 5\n",
      "You entered an odd number\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print (\"Type Control C or -1 to exit\")\n",
    "number = 1\n",
    "while number != -1:\n",
    "    try: \n",
    "        number = int(input(\"Enter a number: \"))\n",
    "        print (\"You entered:\", number)\n",
    "        if number % 2 == 0:\n",
    "            print (\"You entered an even number\")\n",
    "        else:\n",
    "            print (\"You entered an odd number\")\n",
    "    except ValueError:\n",
    "        print (\"That was not a number\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Now when something like “2d” is entered it prints “That was not a number” and continues without exiting.\n",
    "\n",
    "Multiple errors can be caught using more than one except if necessary. In particular, this example will only catch a ValueError but there are many other errors that can be caught e.g.ZeroDivisionError etc.\n",
    "\n",
    "---\n",
    "\n",
    "\n",
    "# Functions\n",
    "\n",
    "---\n",
    "\n",
    "Often when writing a program there are times when you want to use the same piece of code multiple times. An example is a sequence translation program. What if you wanted to translate multiple sequence?\n",
    "\n",
    "This is where functions are used (or subroutines or methods, depending on the language).\n",
    "\n",
    "A function is a block of code that can be used multiple times to perform a particular piece of work.\n",
    "\n",
    "A function is defined by the keyword def followed by the function name. Note the brackets following the name, as functions can optionally take arguments.\n",
    "\n",
    "The structure of a function is like Python control methods, the function name ending in a colon and the scope of the function identified by indented code:\n",
    "\n",
    "<b>def hello():<br>\n",
    "    print (\"Hello World!\")</b><br><br>\n",
    "    \n",
    "\n",
    "A function has to be declared in the program before it can be used.\n",
    "\n",
    "The function is called by simply using its name, and including arguments if required.\n",
    "\n",
    "The code within the function is then called:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello World!\n",
      "Hello World!\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def hello():\n",
    "\n",
    "    print (\"Hello World!\")\n",
    "       \n",
    "hello()\n",
    "hello()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "## Passing Arguments\n",
    "\n",
    "---\n",
    "\n",
    "Any number of arguments can be passed to a function:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "EOL while scanning string literal (<ipython-input-3-35fc2e4ef642>, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-3-35fc2e4ef642>\"\u001b[0;36m, line \u001b[0;32m3\u001b[0m\n\u001b[0;31m    print (\"Hello“, name1, “and”, name2)\u001b[0m\n\u001b[0m                                        ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m EOL while scanning string literal\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def hello(name1, name2):\n",
    "    print (\"Hello“, name1, “and”, name2)\n",
    "       \n",
    "\n",
    "hello(\"Lenka\", \"Ema\")\n",
    "                                                                      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "## Returning Values\n",
    "\n",
    "---\n",
    "\n",
    "Functions can also return data:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    }
   ],
   "source": [
    "def add(num1, num2):\n",
    "    num3 = num1 + num2\n",
    "    return num3\n",
    "\n",
    "\n",
    "num = add(3, 4)\n",
    "print (num)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Multiple values ca be returned and assigned to variables when the function is called:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid character in identifier (<ipython-input-5-68b2711ecb5c>, line 5)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-5-68b2711ecb5c>\"\u001b[0;36m, line \u001b[0;32m5\u001b[0m\n\u001b[0;31m    c = num1 – num2\u001b[0m\n\u001b[0m             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid character in identifier\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def calc(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 – num2\n",
    "\n",
    "    return a,b,c\n",
    "\n",
    "   \n",
    "d, e, f = calc(10, 5)\n",
    "print (d, e, f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "Although the previous version actually returns a tuple it is possible to explicitly return one, or a list:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid character in identifier (<ipython-input-6-143e0fde6c34>, line 5)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-6-143e0fde6c34>\"\u001b[0;36m, line \u001b[0;32m5\u001b[0m\n\u001b[0;31m    c = num1 – num2\u001b[0m\n\u001b[0m             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid character in identifier\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def calcTuple(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 – num2\n",
    "    \n",
    "    return (a,b,c)\n",
    "\n",
    "def calcList(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 – num2\n",
    "\n",
    "return [a,b,c] \n",
    "\n",
    "t = calcTuple(10, 5)\n",
    "\n",
    "print (\"A tuple of values is returned:\")\n",
    "print (t)\n",
    "\n",
    "l = calcList(10, 5)\n",
    "\n",
    "print (\"A list of values is returned:\")\n",
    "print (l)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "## Exercise 14\n",
    "\n",
    "a) Write a script that includes a function that takes 2 numbers, multiplies them together and returns the result. The script should prompt the user for the numbers, pass them to the function and print out the result.\n",
    "\n",
    "b) Create a script that tests the example functions above, which return multiple values, a list and a tuple.\n",
    "\n",
    "Edit the script to demonstrate that it does actually return a list and a tuple.\n",
    "\n",
    "HINT: You can do this by assigning the returned values to a single variable, which will be the tuple, and then printing the index positions.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">here</a>.)\n",
    "\n",
    "---\n",
    "\n",
    "## Passing Other Data as an Argument\n",
    "\n",
    "---\n",
    "\n",
    "Any data type, or object, can be passed to a Python function e.g. list, tuple, dictionary etc:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-7-720756fece7f>, line 24)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-7-720756fece7f>\"\u001b[0;36m, line \u001b[0;32m24\u001b[0m\n\u001b[0;31m    return a,b,c d, e, f = calc(10, 5)\u001b[0m\n\u001b[0m                 ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# Exercise 14a\n",
    "def add(num1, num2): \n",
    "    num3 = num1 + num2\n",
    "    return num3 \n",
    "\n",
    "n1 = int(input(\"Enter first number: \"))\n",
    "n2 = int(input(\"Enter second number: \"))\n",
    "\n",
    "total = add(n1, n2)\n",
    "print (n1, \"+\", n2, \"=\", total)\n",
    "\n",
    "# This code is obviously more verbose than it needs to be,\n",
    "# for instance there is no need for the \"total\" variable\n",
    "# and the print statement could be:\n",
    "#\n",
    "# print (n1, \"+\", n2, \"=\", add(n1, n2))\n",
    "\n",
    "# Exercise 14b\n",
    "# Calc function \n",
    "def calc(num1, num2): \n",
    "    a = num1 + num2 \n",
    "    b = num1 * num2 \n",
    "    c = num1 - num2 \n",
    "    return a,b,c d, e, f = calc(10, 5) \n",
    "print (d, e, f) \n",
    "\n",
    "# Tuple function \n",
    "def calcTuple(num1, num2): \n",
    "    a = num1 + num2 \n",
    "    b = num1 * num2 \n",
    "    c = num1 - num2 \n",
    "    return (a,b,c) \n",
    "\n",
    "tup = calc(10, 5) \n",
    "print (\"First value =\", tup[0]) \n",
    "print (\"Second value =\", tup[1]) \n",
    "print (\"Third value =\", tup[2]) \n",
    "\n",
    "# List function \n",
    "def calcList(num1, num2): \n",
    "    a = num1 + num2 \n",
    "    b = num1 * num2 \n",
    "    c = num1 - num2 \n",
    "    return (a,b,c) \n",
    "lis = calc(10, 5) \n",
    "print (\"First value =\", lis[0]) \n",
    "print (\"Second value =\", lis[1]) \n",
    "print (\"Third value =\", lis[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-8-f9bad00240e0>, line 11)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-8-f9bad00240e0>\"\u001b[0;36m, line \u001b[0;32m11\u001b[0m\n\u001b[0;31m    print calcSum(sumlist)\u001b[0m\n\u001b[0m                ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def calcSum(list1):\n",
    "    total = 0\n",
    "    for num in list1:\n",
    "        total += num\n",
    "\n",
    "    return num\n",
    "\n",
    "\n",
    "sumlist = (1, 2, 3, 4, 5)\n",
    "print calcSum(sumlist)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First value = 30\n",
      "Second value = 200\n",
      "Third value = -10\n",
      "First value = 30\n",
      "Second value = 200\n",
      "Third value = -10\n",
      "Key is  ttt and value is F\n",
      "Key is  tta and value is L\n",
      "Key is  gga and value is G\n",
      "Key is ttt and value is F\n",
      "Key is tta and value is L\n",
      "Key is gga and value is G\n"
     ]
    }
   ],
   "source": [
    "# Exercise 15\n",
    "# CalcTuple Function\n",
    "\n",
    "def calcTuple(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 - num2\n",
    "    return (a,b,c) \n",
    "\n",
    "tup = calcTuple(10, 20)\n",
    "print (\"First value =\", tup[0])\n",
    "print (\"Second value =\", tup[1])\n",
    "print (\"Third value =\", tup[2])\n",
    "\n",
    "\n",
    "# CalcList function\n",
    "def calcList(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 - num2\n",
    "    return [a,b,c] \n",
    "\n",
    "lis = calcTuple(10, 20)\n",
    "print (\"First value =\", lis[0])\n",
    "print (\"Second value =\", lis[1])\n",
    "print (\"Third value =\", lis[2])\n",
    "\n",
    "\n",
    "# Dictionary function\n",
    "# Prints the keys and values of a dictionary\n",
    "\n",
    "def showDictionary(dic):\n",
    "    for k in dic.keys():\n",
    "        print (\"Key is \", k, \"and value is\", dic[k])\n",
    "\n",
    "codons = {'ttt':'F', 'tta':'L', 'gga':'G'}\n",
    "showDictionary(codons)\n",
    "\n",
    "# Dictionary and list function\n",
    "# Prints the keys and values of a dictionary using list of keys\n",
    "\n",
    "def showDictionary(lis, dic):\n",
    "    for k in lis:\n",
    "        print (\"Key is\", k, \"and value is\", dic[k])\n",
    "\n",
    "codon_list = ('ttt', 'tta', 'gga')\t\n",
    "codons = {'ttt':'F', 'tta':'L', 'gga':'G'}\n",
    "showDictionary(codon_list, codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "## Exercise 15\n",
    "\n",
    "---\n",
    "\n",
    "Create a script that tests the 3 example functions above.\n",
    "\n",
    "Edit the script to contain a function that takes a dictionary as an argument and tests it.\n",
    "\n",
    "Edit the script further so that the function takes a list and dictionary as arguments and tests it. A suitable test would be for the list to contain the key values from the dictionary and then iterate the list in the function and print out the associated dictionary values.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">here</a>.)\n",
    "\n",
    "---\n",
    "\n",
    "## Passing Function Arguments by Keywords\n",
    "\n",
    "\n",
    "In the examples above, when multiple arguments are passed to a function, they are associated to parameters by their position. For example, with the following function definition:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;def calcValue(num1, num2, num3):</b><br>\n",
    "\n",
    "If it is called by:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;calcValue(5, 12, 16)</b><br>\n",
    "\n",
    "Then 5 would be assigned to num1, 12 to num2 and 16 to num3, as expected.\n",
    "\n",
    "However, arguments can also be passed by keywords and are then assigned to parameters by explicitly naming them. A simple calculation function could be:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n"
     ]
    }
   ],
   "source": [
    " \n",
    "def calcValue(num1, num2, num3):\n",
    "    total = num1 + num2 - num3\n",
    "   \n",
    "    return total\n",
    "\n",
    "\n",
    "ans = calcValue(5, 10, 6)\n",
    "\n",
    "print (ans)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Alternatively the arguments to the function can be explicitly named, and therefore in any order:\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n",
      "9\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def calcValue(num1, num2, num3):\n",
    "    total = num1 + num2 - num3\n",
    "   \n",
    "    return total\n",
    "\n",
    "ans = calcValue(num1=5, num2=10, num3=6)\n",
    "print (ans)\n",
    "\n",
    "ans = calcValue(num3=6, num2=10, num1=5)\n",
    "print (ans)\n",
    "\n",
    "ans = calcValue(num2=10, num1=5, num3=6)\n",
    "print (ans)\n",
    "\n",
    "# etc ...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "All versions will produce the same answer.\n",
    "\n",
    "An advantage is that you do not have to know in what order parameters are declared in the function.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 16\n",
    "\n",
    "---\n",
    "\n",
    "Create a script that tests the various ways to pass arguments to a function.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">here</a>.)\n",
    "\n",
    "---\n",
    "\n",
    "\n",
    "## Default Values of Parameters\n",
    "\n",
    "---\n",
    "\n",
    "It is possible to set default values of parameters in the function definition:\n",
    " \n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;def calcValue(num1, num2=20, num3=8):</b><br>\n",
    "\n",
    "The parameters num1 and num2 now have default values so the function can be called with just a single value for num3, or with 2 values and num3 will be the default, or with 3 values and both defaults will be overwritten. Finally, one of the default values can be changed if required by specifying the defined name:\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n",
      "9\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "# Exercise 16\n",
    "# Function \n",
    "def calcValue(num1, num2, num3): \n",
    "    total = num1 + num2 - num3 \n",
    "    return total\n",
    "\n",
    "# Call function as normal \n",
    "ans = calcValue(5, 10, 6) \n",
    "print (ans) \n",
    "\n",
    "# Call function with keywords \n",
    "ans = calcValue(num1=5, num2=10, num3=6) \n",
    "print (ans) \n",
    "\n",
    "# Call function with keywords in different order \n",
    "ans = calcValue(num3=6, num2=10, num1=5) \n",
    "print (ans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "must be str, not int",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-20ace3032eb7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# Function can be called with just a single value for num\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mans\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcalcValue\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m6\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"Version 1: \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mans\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;31m# Called with 2 values and num3 will be the default\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: must be str, not int"
     ]
    }
   ],
   "source": [
    "\n",
    "def calcValue(num1, num2=20, num3=8):\n",
    "    total = num1 + num2 - num3\n",
    "\n",
    "    return total\n",
    "\n",
    "# Function can be called with just a single value for num\n",
    "ans = calcValue(6)\n",
    "print (\"Version 1: \" + ans)\n",
    "\n",
    "# Called with 2 values and num3 will be the default\n",
    "ans = calcValue(6, 10)\n",
    "print (\"Version 2: \" + ans)\n",
    "\n",
    "# Called  with 3 values and both defaults will be overwritten\n",
    "ans = calcValue(6, 10, 5)\n",
    "print (\"Version 3: \" + ans)\n",
    "\n",
    "# Default value changed by specifying the defined name\n",
    "ans = calcValue(6, num3=4)\n",
    "print (\"Version 4: \" + ans)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "NOTE: An example of the use of default values is with the Python open file command when opening a file for reading:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;f = open(‘test.txt’, ‘r’)</b><br>\n",
    "\n",
    "The ‘r’ is optional as it is a default value.\n",
    "\n",
    "---\n",
    "\n",
    "## Variable Number of Parameters\n",
    "\n",
    "---\n",
    "\n",
    "A function can take additional optional arguments by prefixing the last parameter with *.\n",
    "\n",
    "Optional arguments are then available in a tuple referenced by this parameter:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def calcValue(num1, num2, *morenums):\n",
    "    total = num1 + num2\n",
    "    for n in morenums:\n",
    "        total += n\n",
    "\n",
    "    return total\n",
    "\n",
    "  \n",
    "ans = calcValue(2, 4,  5, 12, 15)\n",
    "print (ans)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "In this case 2 would be assigned to num1, 4 to num2 and tuple created from the rest (5, 12, 15).\n",
    " \n",
    "If ** is used the arguments are then converted in to a dictionary.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Val1 is First\n",
      "Val2 is Second\n",
      "The dictionary:\n",
      "Key is Third Value is 3\n",
      "Key is Fourth Value is 4\n",
      "Key is Fifth Value is 5\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def testFunc (val1, val2, **morevals):\n",
    "    print (\"Val1 is\", val1)\n",
    "    print (\"Val2 is\", val2)\n",
    "\n",
    "    print (\"The dictionary:\")\n",
    "\n",
    "    for k in morevals.keys():\n",
    "        print (\"Key is\", k, \"Value is\", morevals[k])\n",
    "\n",
    "\n",
    "testFunc(\"First\", \"Second\", Third= 3, Fourth= 4, Fifth= 5)\n",
    "                                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "NOTE: The keys in the dictionary section of the function call do not have quotes.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 17\n",
    "\n",
    "---\n",
    "\n",
    "Create a script that tests the examples given above for default parameters and variable numbers of parameters.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 18\n",
    "\n",
    "---\n",
    "\n",
    "In the Example Program - Sequence Translation part of the tutorial you were provided with the code to translate a nucleotide sequence into amino acids, which you tested in exercise 8.\n",
    "\n",
    "The fasta file (<b>sequence.txt</b>) and codons file (<b>codons.txt</b>) needed to test the script can be downloaded from the <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">Exercise Answers</a> page.\n",
    "\n",
    "Modify the code so that it translates the sequence in all 6 reading frames. You should include a function that carries out the translation and call it 6 times.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 19\n",
    "\n",
    "---\n",
    "\n",
    "Modify your code from the previous exercise so that the function translates only a section of the sequence. It should also only translate the sequence in a single frame, defined as a function parameter.\n",
    "\n",
    "\n",
    "For example, the function could take 4 arguments:\n",
    "\n",
    "    Sequence\n",
    "    Start\n",
    "    End\n",
    "    Sequence strand (1 or -1)\n",
    "\n",
    "Remember, for the minus strand you will need to complement the sequence.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 20\n",
    "\n",
    "---\n",
    "\n",
    "Modify your code from the previous exercise so that the function extracts  multiple sections of the sequence and translates them. The previous example would only translate a single exon gene but this version should translate a multi exon gene.\n",
    "\n",
    "The number of exons will vary so your function will have to take a single list of start and end positions.\n",
    "\n",
    "To test this version of the function you can use a different and much larger sequence (<b>blumeria_seq.fasta</b>), available from the <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">Exercise Answers</a> page. This is a contig annotation from Blumeria graminis and the genes in the sequence are annotated so you can use those positions to test your function. There are 3 genes, 2 on the forward strand and 1 on the reverse:\n",
    "\n",
    "Gene 1\n",
    "\n",
    "Strand +\n",
    "Exon 1     Start     4550     End     4636           \n",
    "Exon 2     Start     4698     End    4796       \n",
    "Exon 3     Start     4852     End    5565   \n",
    "Exon 4     Start     5623     End     5715   \n",
    "\n",
    "Gene 2\n",
    "\n",
    "Strand +\n",
    "Exon 1     Start      7647       End      7759   \n",
    "Exon 2     Start      7854      End      8216   \n",
    "Exon 3     Start      8267      End      8660   \n",
    "\n",
    "Gene 3\n",
    "\n",
    "Strand –\n",
    "Exon 1     Start    17433    End      17709\n",
    "Exon 2     Start      17243  End      17370       \n",
    "\n",
    "You could also change the function to use arguments with default values. For example, a default could be “strand=1”, so the function assumes the positive strand.\n",
    "\n",
    "NOTE: The correct translations are:\n",
    "\n",
    "Gene 1 Translation: MSLPTVLLVYKIIHTVDEWNDLSRIATLKVFNGNSRQDFLKSCEDGEFNEVTIIYYSNESVRLIGQLDQELLFKLPPSIRYICHHGAGYDSIDIAACTERSIQVSHTPIAVNKSTADMTLFLILGALRRIHVPYLAVRAGQWRGSTQLGHDPENKLLGILGMGGIGQEVAKRAKAFGMKVQYHNRSRLVPELEQGADYVSFEHLIKTSDILSLNCSLNKATIGIIGKDELAQMKQGVIVVNTARGKLIDEFALVKALESGKVFSAGLDVFEKEPSIDDTLLSSPNVVLTPHIGTATVETQRAMELLVLQNIENALKTDALVTQVQEQIQA*\n",
    "Gene 2 Translation: MSVVSLLGVNVLQNPARFGDPYEFEITFECLETLQKGTYIVGHKLTPVFNKLIFNSNDHDQELDSLLVGPIPVGVNKFIFVADPPDTNKIPDAEILGVTVILLTCAYDGREFVRVGYYVNNEYDSDELNTDPPAKPILEKVRRNILAEKPRVTRFAIKWDSDDSAPPLYPPEQPEADLVADGEEYGAEEAEDEEEEESADGPEVPADPDVMIEDSEVAGAMVETVKATEEESDAGSEDLEAESSGSEEDEIEEDEEREDEPEEAMDLDGAGKPNGATSSTHNTDTTMAH*\n",
    "Gene 3 Translation: MWSLQYLALLVFVSRAAANYQHWDIDSAVNCQNNIWTSNYLKQVRLRYCNHPPSPQSLVDISRHPNSQLHSNRSTNIYFLSIPRPGPDGELGDDVSNYYMLVDADCNYYSVVGLNARYAQGFVLNSQTVPCRLA*\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">here</a>.)\n",
    "\n",
    "---\n",
    "\n",
    "# Producing Plots\n",
    "\n",
    "---\n",
    "\n",
    "<b>IMPORTANT - To test the code examples provided in this section of the tutorial you will need to copy them to a Linux text editor to create a script that can be run. This is because the code needs access to the matplotlib module.</b>\n",
    "\n",
    "---\n",
    "\n",
    "The python module <b>matplotlib</b> can be used to generate plots and is available in the <b>SciPy</b> package.\n",
    "\n",
    "This package should be installed but to test it and to demonstrate the functionality of matplotlib create the following script and run it: \n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from pylab import *<br><br>\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;x_data = [0, 1, 2,  3, 4, 5, 6]<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;y_data = [0, 1, 0, -1, 0, 1, 0]<br><br>\n",
    "   \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;plot(x_data, y_data)<br><br>\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;xlabel(\"This is the x label\")<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ylabel(\"This is the y label\")<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title(\"Here is the title\")<br><br>\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;grid(True)<br><br>\n",
    "   \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show()</b><br><br>\n",
    "\n",
    "The output should look like this:\n",
    "\n",
    "![title](img/plot1.jpg)\n",
    "\n",
    "### Explanation of the Code\n",
    "\n",
    "<b>from pylab import *</b> - imports all of the modules from the pylab library\n",
    "\n",
    "<b>x_data = [0, 1, 2,  3, 4, 5, 6]</b>\n",
    "<b>y_data = [0, 1, 0, -1, 0, 1, 0]</b> - these are the lists for the X and Y axis coordinates\n",
    "   \n",
    "<b>plot(x_data, y_data)</b> - plots the graph using coordinates declared above\n",
    "\n",
    "<b>xlabel(\"This is the x label\")</b>\n",
    "<b>ylabel(\"This is the y label\")</b>\n",
    "<b>title(\"Here is the title\")</b> - labels for the X and Y axis and the title\n",
    "\n",
    "<b>grid(True)</b> - shows the grid on the plot\n",
    "   \n",
    "<b>show()</b> - displays the plot on the screen\n",
    "\n",
    "If you want to save the image rather than displaying it you can use the savefig command rather than show:\n",
    "\n",
    "<b>savefig('plot1.png', dpi=80)</b>\n",
    "\n",
    "---\n",
    "\n",
    "## Hydrophobicity Plot\n",
    "\n",
    "---\n",
    "\n",
    "Every amino acid has a certain hydrohobicity prevalence and this can be used to predict membrane spanning regions of proteins as these regions will have high values. We can use these values to plot the hydrophobicity of a protein along its entire length. As an example write a script for the code below and run it. This code uses the Kyte and Doolittle hydrophobicity scale, which is widely used to delineate the hydrophobic properties of proteins:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from pylab import *<br><br>\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;y_data = [1.8, -3.5, 4.5, -0.7, -0.4, -4.5, -1.6, -3.5, -0.9, 4.5, -0.9, 3.8, 1.8, 3.8,  -0.4, -0.7, 1.8, 3.8, 1.9, -0.4]<br><br>\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;x_data = range(1, len(y_data)+1)<br><br>\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;plot(x_data, y_data)<br><br>\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;xlabel(\"Residue number\")<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ylabel(\"Hydrophobicity\")<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title(\"Kyte & Doolittle Hydrophobicity\")<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;show()</b><br><br>\n",
    "\n",
    "The plot should look like this:\n",
    "\n",
    "![title](img/plot2.jpg)\n",
    "\n",
    "The code is essentially the same as before except for setting the X axis value:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;x_data = range(1, len(y_data)+1)</b>\n",
    "\n",
    "This is for the number of amino acid residues so is equal to the number of values plotted for the Y axis.\n",
    "\n",
    "Normally a range starts at 0 but protein amino acids coordinates start at 1 so the range is adjusted accordingly so the first point on the plot is at residue 1.\n",
    "\n",
    "The adjustment to the range means that there is a gap at the start of the plot between residue 0 and 1. this can be corrected by modifying the axis, in this case the X axis. Add the following code to your script, after <b>plot(x_data, y_data)</b>, and run it again:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;residue_count = len(y_data)\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;axis(xmin = 1, xmax = residue_count)</b> - adjusts the X axis to start at 1\n",
    "\n",
    "\n",
    "The gap should now be removed.\n",
    "\n",
    "The Y axis be adjusted using the same axis command.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 21\n",
    "\n",
    "a) The exercise is to produce a plot for a real protein using Kyte and Doolittle values. For this task you will need the following files:\n",
    "\n",
    "5HT1A_HUMAN.fasta - This is the human 5-hydroxytryptamine receptor 1A protein sequence that you will use for the plot.\n",
    "\n",
    "hydrophobicity_values.txt - This is a file of the Kyte and Doolittle values for all amino acids. The file has the amino acid on one line and the hydrophobicity value on the next:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.8<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;R<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-4.5<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;etc</b><br><br>\n",
    "\n",
    "These can be downloaded from the <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">Exercise Answers</a> page:\n",
    "\n",
    "Write a script with the following functionality:\n",
    "\n",
    "1) Reads the 5HT1A_HUMAN.fasta file and stores the amino acid sequence and description in separate variables. The file is in fasta format so the first line is just the sequence description, starting with a \">\", and the subsequent lines are the actual sequence. You'll need the description for the plot label so you can use any part of the description line for this.\n",
    "\n",
    "Note that the sequence itself is over multiple lines.\n",
    "\n",
    "2) Reads the hydrophobicity_values.txt file and stores the values in a dictionary. The keys for the dictionary should be the amino acids and the values the hydrophobicity values.\n",
    "\n",
    "3) Test the dictionary by printing to screen the Kyte and Doolitle hydrophbicity values for each amino acid in the 5HT1A_HUMAN sequence, separated by commas. The output should look like this:\n",
    "\n",
    "1.9, -3.5, 4.2, 3.8, -0.8, -1.6, -0.4, -3.5, -0.4, -3.5, -3.5, -0.7, -0.7, -0.8, -1.6, -1.6, 1.8, -1.6, 2.8, -3.5, -0.7, -0.4, -0.4, -3.5, -0.7, -0.7, -0.4, 4.5  , -0.8, -3.5, 4.2, -0.7, 4.2, -0.8, -1.3, -3.5, 4.2, 4.5, -0.7, -0.8, 3.8, 3.8, 3.8, -0.4, -0.7, 3.8, 4.5, 2.8, 2.5, 1.8, 4.2, 3.8, -0.4, -3.5, 1.8, 1.8, 3.8, -1.3, -3.5, 4.2, 3.8, -3.5, -3.9, -0.9, -0.7, 3.8, -0.4, -3.5, 4.2, -0.7, 2.5, -3.5, etc ...\n",
    "\n",
    "Modify the script using the earlier code to draw the hydrophobicity plot for the 5HT1A_HUMAN sequence.\n",
    "\n",
    "The output should look like this:\n",
    "\n",
    "![title](img/plot3.jpg)\n",
    "\n",
    "The plot produced is difficult to interpret as it’s very noisy. The reason for this is that each residue has a hydrophobicity value but for hydrophobic regions, such as trans-membrane, several hydrophobic residues in a row are more significant.\n",
    "\n",
    "To better predict these regions it is better to take an average for a residues and its neighbours. To start, try averaging the residue and its 2 neighbours, left and right. As the first and last residues don’t have neighbours they’ll be ignored.\n",
    "\n",
    "b) Modify your code to plot the average hydrophobicity values for 3 amino acids at a time. When calculating the average you may have to ensure all of the variables are numbers and not strings. You can do that with:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;average = float(var1) + float(var2) + float(var3) / 3</b><br><br>\n",
    "   \n",
    "\n",
    "The resulting plot should look like this:\n",
    "\n",
    "![title](img/plot4.jpg)\n",
    "\n",
    "This is called <b>data smoothing</b> as it filters out the noise to better display the underlying pattern.\n",
    "\n",
    "The code currently averages over 3 residues but when trying to identify transmembrane regions a window size of 15 is more suitable.\n",
    "\n",
    "c) Update your code to implement this value, which will require the average value of each residue and its 7 neighbours either side.\n",
    "\n",
    "It would be inefficient to create different variables for all 15 residue values, such as left1, left 2, etc, so calculate the average for each 15 residue group by using a loop.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://web.bioinformatics.ic.ac.uk/msc/pythontutorial/exercises.html\">here</a>.)\n",
    "\n",
    "---\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n",
      "8\n",
      "11\n",
      "22\n",
      "38\n",
      "Val1 is First\n",
      "Val2 is Second\n",
      "The dictionary\n",
      "Key is Third Value is 3\n",
      "Key is Fourth Value is 4\n",
      "Key is Fifth Value is 5\n"
     ]
    }
   ],
   "source": [
    "# Exercise 17\n",
    "\n",
    "# Default parameters\n",
    "def calcValue(num1, num2=20, num3=8):\n",
    "    total = num1 + num2 - num3\n",
    "    return total\n",
    "ans = calcValue(6)\n",
    "print (ans)\n",
    "\n",
    "# Call it with 2 numbers so only num3 uses the default\n",
    "ans = calcValue(6, 10)\n",
    "print (ans)\n",
    "\n",
    "# Can still call it with 3 different numbers\n",
    "ans = calcValue(6, 10, 5)\n",
    "print (ans)\n",
    "\n",
    "# Can also change just one of the defaults\n",
    "ans = calcValue(6, num3=4)\n",
    "print (ans)\n",
    "\n",
    "# Variable number of parameters in a list\n",
    "def calcValue(num1, num2, *morenums):\n",
    "    total = num1 + num2\n",
    "    for n in morenums:\n",
    "        total += n\n",
    "    return total\n",
    "\n",
    "ans = calcValue(2, 4,  5, 12, 15)\n",
    "print (ans)\n",
    "\n",
    "# Variable number of parameters in a dictionary\n",
    "def testFunc (val1, val2, **morevals):\n",
    "    print (\"Val1 is\", val1)\n",
    "    print (\"Val2 is\", val2)\n",
    "    print (\"The dictionary\")\n",
    "    for k in morevals.keys():\n",
    "        print (\"Key is\", k, \"Value is\", morevals[k])\n",
    "\n",
    "testFunc(\"First\", \"Second\", Third=3, Fourth=4, Fifth=5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward Strand\n",
      "Frame 1 :\n",
      " MDLEKNYPTPRTSRTGHGGVNQLGGVFVNGRPLPDVVRQRIVELAHQGVRPCDISRQLRVSHGCVSKILGRYYETGSIKPGVIGGSKPKVATPKVVEKIAEYKRQNPTMFAWEIRGRLLAERVCDNDTVPSVSSINRIIRTKVQQPPNQPVPASSHSIVSTGSVTQVSSVSTDSAGSSYSISGILGITSPSADTNKRKRDEGIQESPVPNGYSLPGRDFLRKQMRGDLFTQQQLEVLDRVFERQHYSDIFTTTEPIKPEQTTEYSAMASLAGGLDDMKANLASPTPADIGSSVPGPQSYPIVTGRDLASTTLPGYPPHVPPAGQGSYSAPTLTGMVPGSEFSGSPYSHPQYSSYNDSWRFPNPGLLGSPYYYSAAARGAAPPAAATAYDRH-\n",
      "Frame 2 :\n",
      " WI*RKIIRLLGPAGQDMEE*ISLGGFL*MDGHSRM*SARG*WNLLIKVSGPATSPGSFGSAMVVSAKFLAGIMRQEASSLG*LEDPNQRSPHPKWWKKSLSINAKIPPCLPGRSGAGCWQSGCVTMTPCLASVPSTGSSGQKYSSHPTNQSQLPVTA*CPLAP*RRCPR*ARIRPARRTPSAASWASRPPAPTPTSARETKVFRSLRCRTATRFRAETSSGSRCGETCSHSSSWRCWTACLRGSTTQTSSPPQSPSSPSRPQSIQPWPRWLVGWTT*RPIWPAPPLLTSGAVCQARSPTPL*QAVTWRARPSPGTLHTSPPLDRAATQHRR*QGWCLGVSFPGVPTATLSIPRTTTPGGSPTRGCLAPPTIIALPPEEPPHLQPPLPMTVT-\n",
      "Frame 3 :\n",
      " GFREKLSDSSDQQDRTWRSESAWGGFCEWTATPGCSPPEDSGTCSSRCQALRHLQAASGQPWLCQQNSWQVL*DRKHQAWGNWRIQTKGRHTQSGGKNR*V*TPKSHHVCLGDQGPAAGRAGV*Q*HRA*RQFHQQDHPDKSTAATQPTSPSFQSQHSVHWLRDAGVLGEHGFGRLVVLHQRHPGHHVPQRRHQQAQERRRYSGVSGAERLLASGQRLPPEADAGRLVHTAAAGGAGPRV*EAALLRHLHHHRAHQARADHRVFSHGLAGWWAGRHEGQSGQPHPC*HREQCARPAVLPHCDRP*LGEHDPPRVPSTRPPRWTGQLLSTDADRDGAWE*VFRESLQPPSVFLVQRLLEVPQPGAAWLPLLL*RCRPRSRPTCSRHCL*PSL-\n",
      "\n",
      "Reverse Strand\n",
      "Frame 1 :\n",
      " -VTVIGSGGCRWGGSSGGSAIIVGGAKQPRVGEPPGVVVRGILRVAVGTPGKLTPRHHPCQRRC*VAALSSGGDVWRVPGEGRARQVTACHNGVGLRAWHTAPDVSRGGAGQIGLHVVQPTSQRGHG*ILCGLLGLDGLCGGEDV*VVLPLKHAVQHLQLLLCEQVSPHLLPEEVSARKRVAVRHRRLLNTFVSLALVGVGAGGRDAQDAADGVRRAGRIRAHRGHLRHGASGHYAVTGSWDWLVGWLLYFCPDDPVDGTDARHGVIVTHPLCQQPAPDLPGKHGGILAFILSDFFHHFGCGDLWFGSSNYPRLDASCLIIPAKNFADTTMADPKLPGDVAGPDTLMSKFHYPLADYIREWPSIHKNPPKLIHSSMSCPAGPRSRIIFL*IH\n",
      "Frame 2 :\n",
      " Q*RS*AVAAAGGAAPRAAAL***GEPSSPGLGNLQESLYEEY*GWL*GLPENSLPGTIPVSVGAE*LPCPAGGTCGGYPGRVVLAKSRPVTMG*DCGPGTLLPMSAGVGLARLAFMSSSPPASEAMAEYSVVCSGLMGSVVVKMSE*CCLSNTRSSTSSCCCVNKSPRICFRRKSLPGSE*PFGTGDS*IPSSLLRLLVSALGDVMPRMPLMEYDEPAESVLTEDTCVTEPVDTML*LEAGTGWLGGCCTFVRMILLMELTLGTVSLSHTRSASSRPLISQANMVGFWRLYSAIFSTTLGVATFGLDPPITPGLMLPVS*YLPRILLTQPWLTRSCLEMSQGLTP**ASSTILWRTTSGSGRPFTKTPPS*FTPPCPVLLVRGVG*FFSKS-\n",
      "Frame 3 :\n",
      " SDGHRQWRLQVGRLLGRQRYNSRGSQAAPGWGTSRSRCTRNTEGGCRDSRKTHSQAPSLSASVLSSCPVQRGGRVEGTRGGSCSPSHGLSQWGRTAGLAHCSRCQQGWGWPDWPSCRPAHQPARPWLNTLWSARA*WALWW*RCLSSAASQTRGPAPPAAAV*TSLPASASGGSLCPEASSRSAPETPEYLRLSCACWCRRWGT*CPGCR*WSTTSRPNPCSPRTPASRSQWTLCCDWKLGLVGWVAAVLLSG*SC*WN*R*ARCHCHTPALPAAGP*SPRQTWWDFGVYTQRFFPPLWVWRPLVWILQLPQA*CFLSHNTCQEFC*HNHG*PEAAWRCRRA*HLDEQVPLSSGGLHPGVAVHSQKPPQADSLLHVLSCWSEESDNFSLNP-\n"
     ]
    }
   ],
   "source": [
    "# Exercise 18\n",
    "\n",
    "# Dictionary for the reverse complementing of the nucleic acids\n",
    "trans = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "def translate(seq):\n",
    "\taaseq = \"\"\n",
    "\tfor count in range(0, len(seq), 3):  # Work through the list\n",
    "\t\tcodon = seq[count:count+3]   # Get 3 nucleotides\n",
    "\t\tif codon in codons:\n",
    "\t\t\taa = codons[codon]   # Get the associated aa\n",
    "\t\telse:\n",
    "\t\t\taa = '-'\n",
    "\t\t\n",
    "\t\taaseq += aa\n",
    "\n",
    "\treturn aaseq\n",
    "\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "\tcodons_list = codons_file.readlines()\n",
    "\n",
    "\tcodons = {}  # Initialise the dictionary\n",
    "\tfor count in range(0, len(codons_list), 2):  # Work through the list\n",
    "\t\tkey = codons_list[count].rstrip()   # Get the codon\n",
    "\t\tkey = key.upper()\n",
    "\t\tvalue = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "\t\tvalue = value.upper()\t\n",
    "\t\tcodons[key] = value  # Add to dictionary\n",
    "\n",
    "with open('sequence.txt') as in_file:\n",
    "\n",
    "\tseq_list = in_file.readlines()  # Read the file to a list\n",
    "\tseq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "\tseq = seq_list.pop(0)  # Initialise the sequence\n",
    "\tseq = seq.rstrip()  # Remove the newline character\n",
    "\t\n",
    "\tfor line in seq_list:  #  Append the rest of the sequence to seq\n",
    "\t\tseq += line.rstrip()\n",
    "\tseq = seq.upper()  # Upper case the sequence\n",
    "\n",
    "\tfor i in range(1, 3):\n",
    "\t\tif i == 1:\n",
    "\t\t\tprint (\"Forward Strand\")\n",
    "\t\telse:\n",
    "\t\t\tprint (\"\\nReverse Strand\")\n",
    "\t\tfor j in range(1, 4):\n",
    "\t\t\taaseq = translate(seq[j-1:])\n",
    "\n",
    "\t\t\tprint (\"Frame\", j, \":\\n\", aaseq)\n",
    "\t\t\n",
    "\t\t# Reverse the sequence\n",
    "\t\trevseq = \"\"\n",
    "\t\tseq = seq.upper()\n",
    "\t\tif i == 1:\n",
    "\t\t\tseq = seq[::-1]\n",
    "\t\t\tfor base in seq:\n",
    "\t\t\t\tif base in trans:\n",
    "\t\t\t\t\trevseq += trans[base]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\trevseq += \"N\"\n",
    "\t\t\tseq = revseq\n",
    "\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LHPGVAVHSQKPPQADSLLHVLSCWSEESDNFS\n"
     ]
    }
   ],
   "source": [
    "# Exercise 19\n",
    "# Dictionary for the reverse complementing of the nucleic acids\n",
    "\n",
    "trans = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "\n",
    "def translate(seq, start, end, strand):\n",
    "\n",
    "\tseq = seq[start:end]\n",
    "\n",
    "\tif strand == -1:\n",
    "\t\t# Reverse the sequence\n",
    "\t\trevseq = \"\"\n",
    "\t\t\n",
    "\t\tseq = seq[::-1]\n",
    "\t\tfor base in seq:\n",
    "\t\t\tif base in trans:\n",
    "\t\t\t\trevseq += trans[base]\n",
    "\t\t\telse:\n",
    "\t\t\t\trevseq += \"N\"\n",
    "\t\tseq = revseq\n",
    "\n",
    "\taaseq = \"\"\n",
    "\tfor count in range(0, len(seq), 3):  # Work through the list\n",
    "\t\tcodon = seq[count:count+3]   # Get 3 nucleotides\n",
    "\t\tif codon in codons:\n",
    "\t\t\taa = codons[codon]   # Get the associated aa\n",
    "\t\telse:\n",
    "\t\t\taa = '-'\n",
    "\t\t\n",
    "\t\taaseq += aa\n",
    "\n",
    "\treturn aaseq\n",
    "\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "\tcodons_list = codons_file.readlines()\n",
    "\n",
    "\tcodons = {}  # Initialise the dictionary\n",
    "\tfor count in range(0, len(codons_list), 2):  # Work through the list\n",
    "\t\tkey = codons_list[count].rstrip()   # Get the codon\n",
    "\t\tkey = key.upper()\n",
    "\t\tvalue = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "\t\tvalue = value.upper()\t\n",
    "\t\tcodons[key] = value  # Add to dictionary\n",
    "\n",
    "with open('sequence.txt') as in_file:\n",
    "\n",
    "\tseq_list = in_file.readlines()  # Read the file to a list\n",
    "\tseq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "\tseq = seq_list.pop(0)  # Initialise the sequence\n",
    "\tseq = seq.rstrip()  # Remove the newline character\n",
    "\t\n",
    "\tfor line in seq_list:  #  Append the rest of the sequence to seq\n",
    "\t\tseq += line.rstrip()\n",
    "\tseq = seq.upper()  # Upper case the sequence\n",
    "\n",
    "\taaseq = translate(seq, 10, 109, -1)\n",
    "\n",
    "\tprint (aaseq)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'blumeria_seq.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-dd97dc9710b9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     52\u001b[0m                 \u001b[0mcodons\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m  \u001b[0;31m# Add to dictionary\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 54\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'blumeria_seq.fasta'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0min_file\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m         \u001b[0mseq_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0min_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Read the file to a list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'blumeria_seq.fasta'"
     ]
    }
   ],
   "source": [
    "# Exercise 20 -- version 1\n",
    "# Dictionary for the reverse complementing of the nucleic acids\n",
    "\n",
    "trans = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "\n",
    "def translate(seq, strand, *pos):\n",
    "\n",
    "\tcode_seq = \"\"\n",
    "\tfor p in range(0, len(pos), 2):\n",
    "\t\tstart = pos[p]\n",
    "\t\tend = pos[p+1]\n",
    "\n",
    "\t\tex_seq = seq[start-1:end]\n",
    "\n",
    "\t\tif strand == -1:\n",
    "\t\t\t# Reverse the sequence\n",
    "\t\t\trevseq = \"\"\n",
    "\t\t\n",
    "\t\t\tex_seq = ex_seq[::-1]\n",
    "\t\t\tfor base in ex_seq:\n",
    "\t\t\t\tif base in trans:\n",
    "\t\t\t\t\trevseq += trans[base]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\trevseq += \"N\"\n",
    "\t\t\tex_seq = revseq\n",
    "\t\t\n",
    "\t\tcode_seq += ex_seq\n",
    "\n",
    "\taaseq = \"\"\n",
    "\tfor count in range(0, len(code_seq), 3):  # Work through the list\n",
    "\t\tcodon = code_seq[count:count+3]   # Get 3 nucleotides\n",
    "\t\tif codon in codons:\n",
    "\t\t\taa = codons[codon]   # Get the associated aa\n",
    "\t\telse:\n",
    "\t\t\taa = '-'\n",
    "\t\n",
    "\t\taaseq += aa\n",
    "\n",
    "\treturn aaseq\n",
    "\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "\tcodons_list = codons_file.readlines()\n",
    "\n",
    "\tcodons = {}  # Initialise the dictionary\n",
    "\tfor count in range(0, len(codons_list), 2):  # Work through the list\n",
    "\t\tkey = codons_list[count].rstrip()   # Get the codon\n",
    "\t\tkey = key.upper()\n",
    "\t\tvalue = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "\t\tvalue = value.upper()\t\n",
    "\t\tcodons[key] = value  # Add to dictionary\n",
    "\n",
    "with open('blumeria_seq.fasta') as in_file:\n",
    "\n",
    "\tseq_list = in_file.readlines()  # Read the file to a list\n",
    "\tseq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "\tseq = seq_list.pop(0)  # Initialise the sequence\n",
    "\tseq = seq.rstrip()  # Remove the newline character\n",
    "\t\n",
    "\tfor line in seq_list:  #  Append the rest of the sequence to seq\n",
    "\t\tseq += line.rstrip()\n",
    "\tseq = seq.upper()  # Upper case the sequence\n",
    "\n",
    "\taaseq = translate(seq, 1, 4550, 4636, 4698, 4796, 4852, 5565, 5623, 5715)\n",
    "\n",
    "\tprint (\"Gene 1 Translation:\", aaseq)\n",
    " \n",
    "\taaseq = translate(seq, 1, 7647, 7759, 7854, 8216, 8267, 8660)\n",
    " \n",
    "\tprint (\"Gene 2 Translation:\", aaseq)\n",
    "\n",
    "\taaseq = translate(seq, -1, 17433, 17709, 17243, 17370)\n",
    "\n",
    "\tprint (\"Gene 3 Translation:\", aaseq)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'blumeria_seq.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-9900a893d9aa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     54\u001b[0m                 \u001b[0mcodons\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m  \u001b[0;31m# Add to dictionary\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 56\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'blumeria_seq.fasta'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0min_file\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m         \u001b[0mseq_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0min_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Read the file to a list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'blumeria_seq.fasta'"
     ]
    }
   ],
   "source": [
    "# Exercise 20 -- version 2\n",
    "# Dictionary for the reverse complementing of the nucleic acids\n",
    "\n",
    "trans = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "# As a default value is used in the function definition this time the positions\n",
    "# have to be in a list. The default value also needs to be last\n",
    "\n",
    "def translate(seq, pos, strand=1):\n",
    "\n",
    "\tcode_seq = \"\"\n",
    "\tfor p in range(0, len(pos), 2):\n",
    "\t\tstart = pos[p]\n",
    "\t\tend = pos[p+1]\n",
    "\n",
    "\t\tex_seq = seq[start-1:end]\n",
    "\n",
    "\t\tif strand == -1:\n",
    "\t\t\t# Reverse the sequence\n",
    "\t\t\trevseq = \"\"\n",
    "\t\t\n",
    "\t\t\tex_seq = ex_seq[::-1]\n",
    "\t\t\tfor base in ex_seq:\n",
    "\t\t\t\tif base in trans:\n",
    "\t\t\t\t\trevseq += trans[base]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\trevseq += \"N\"\n",
    "\t\t\tex_seq = revseq\n",
    "\t\t\n",
    "\t\tcode_seq += ex_seq\n",
    "\n",
    "\taaseq = \"\"\n",
    "\tfor count in range(0, len(code_seq), 3):  # Work through the list\n",
    "\t\tcodon = code_seq[count:count+3]   # Get 3 nucleotides\n",
    "\t\tif codon in codons:\n",
    "\t\t\taa = codons[codon]   # Get the associated aa\n",
    "\t\telse:\n",
    "\t\t\taa = '-'\n",
    "\t\n",
    "\t\taaseq += aa\n",
    "\n",
    "\treturn aaseq\n",
    "\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "\tcodons_list = codons_file.readlines()\n",
    "\n",
    "\tcodons = {}  # Initialise the dictionary\n",
    "\tfor count in range(0, len(codons_list), 2):  # Work through the list\n",
    "\t\tkey = codons_list[count].rstrip()   # Get the codon\n",
    "\t\tkey = key.upper()\n",
    "\t\tvalue = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "\t\tvalue = value.upper()\t\n",
    "\t\tcodons[key] = value  # Add to dictionary\n",
    "\n",
    "with open('blumeria_seq.fasta') as in_file:\n",
    "\n",
    "\tseq_list = in_file.readlines()  # Read the file to a list\n",
    "\tseq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "\tseq = seq_list.pop(0)  # Initialise the sequence\n",
    "\tseq = seq.rstrip()  # Remove the newline character\n",
    "\t\n",
    "\tfor line in seq_list:  #  Append the rest of the sequence to seq\n",
    "\t\tseq += line.rstrip()\n",
    "\tseq = seq.upper()  # Upper case the sequence\n",
    "\n",
    "\t# As a default value is used in the function definition this time the positions\n",
    "\t# have to be in a list\n",
    "\t# In the first 2 genes the strand is not included as the default is used\n",
    "\n",
    "\texons = (4550, 4636, 4698, 4796, 4852, 5565, 5623, 5715)\n",
    "\n",
    "\taaseq = translate(seq, exons)\n",
    "\n",
    "\tprint (\"Gene 1 Translation:\", aaseq)\n",
    "\n",
    "\texons = (7647, 7759, 7854, 8216, 8267, 8660)\n",
    " \n",
    "\taaseq = translate(seq, exons)\n",
    " \n",
    "\tprint (\"Gene 2 Translation:\", aaseq)\n",
    "\n",
    "\texons = (17433, 17709, 17243, 17370)\n",
    "\n",
    "\t# For this gene the strand is included to override the default\n",
    "\n",
    "\taaseq = translate(seq, exons, -1)\n",
    "\n",
    "\tprint (\"Gene 3 Translation:\", aaseq)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Exercise 21a\n",
    "from pylab import *\n",
    "\n",
    "in_file = open(\"5HT1A_HUMAN.fasta\")\n",
    "\n",
    "seq = \"\"\n",
    "for line in in_file:\n",
    "\tstrip_line = line.rstrip()\n",
    "\tif(strip_line.startswith('>')):\n",
    "\t\tseqid = strip_line\n",
    "\t\tseqid = seqid.replace('>','')\n",
    "\telse:\n",
    "\t\tseq += strip_line\n",
    "in_file.close()\n",
    "\n",
    "in_file = open(\"hydrophobicity_values.txt\")\n",
    "vals_lines = in_file.readlines()\n",
    "\n",
    "vals = {}  # Initialise the dictionary\n",
    "for count in range(0, len(vals_lines), 2): \n",
    "\tvals[vals_lines[count].rstrip()] = vals_lines[count+1].rstrip()\t\n",
    "\n",
    "in_file.close()\n",
    "\n",
    "y_data = []\n",
    "for count in range(0, len(seq)):  \n",
    "\taa = seq[count]   \n",
    "\tprint (vals[aa]+\",\", end=\"\")\n",
    "\ty_data.append(vals[aa])\n",
    "\n",
    "x_data = range(1, len(y_data)+1)\n",
    "\n",
    "plot(x_data, y_data)\n",
    "\n",
    "xlabel(\"Residue number\")\n",
    "ylabel(\"Hydrophobicity\")\n",
    "title(\"Hydrophobicity Plot for \"+seqid)\n",
    "show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '5HT1A_HUMAN.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-0f40bdc27fb2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpylab\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0min_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"5HT1A_HUMAN.fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '5HT1A_HUMAN.fasta'"
     ]
    }
   ],
   "source": [
    "# Exercise 21b\n",
    "from pylab import *\n",
    "\n",
    "in_file = open(\"5HT1A_HUMAN.fasta\")\n",
    "\n",
    "seq = \"\"\n",
    "for line in in_file:\n",
    "\tstrip_line = line.rstrip()\n",
    "\tif(strip_line.startswith('>')):\n",
    "\t\tseqid = strip_line\n",
    "\t\tseqid = seqid.replace('>','')\n",
    "\telse:\n",
    "\t\tseq += strip_line\n",
    "in_file.close()\n",
    "\n",
    "in_file = open(\"hydrophobicity_values.txt\")\n",
    "vals_lines = in_file.readlines()\n",
    "\n",
    "vals = {}  # Initialise the dictionary\n",
    "for count in range(0, len(vals_lines), 2): \n",
    "\tvals[vals_lines[count].rstrip()] = vals_lines[count+1].rstrip()\t\n",
    "\n",
    "in_file.close()\n",
    "\n",
    "y_data = []\n",
    "for count in range(1, len(seq)-1):  \n",
    "\tleft_aa = seq[count-1]   \n",
    "\tcenter_aa = seq[count]   \n",
    "\tright_aa = seq[count+1]   \n",
    "\taverage = (float(vals[left_aa]) + float(vals[center_aa]) + float(vals[right_aa])) / 3 \n",
    "\ty_data.append(average)\n",
    "\n",
    "\n",
    "x_data = range(1, len(y_data)+1)\n",
    "\n",
    "plot(x_data, y_data)\n",
    "\n",
    "xlabel(\"Residue number\")\n",
    "ylabel(\"Hydrophobicity\")\n",
    "title(\"Hydrophobicity Plot for \"+seqid)\n",
    "show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '5HT1A_HUMAN.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-7c8f2fa6f61f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpylab\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0min_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"5HT1A_HUMAN.fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '5HT1A_HUMAN.fasta'"
     ]
    }
   ],
   "source": [
    "# Exercise 21c\n",
    "from pylab import *\n",
    "\n",
    "in_file = open(\"5HT1A_HUMAN.fasta\")\n",
    "\n",
    "seq = \"\"\n",
    "for line in in_file:\n",
    "\tstrip_line = line.rstrip()\n",
    "\tif(strip_line.startswith('>')):\n",
    "\t\tseqid = strip_line\n",
    "\t\tseqid = seqid.replace('>','')\n",
    "\telse:\n",
    "\t\tseq += strip_line\n",
    "in_file.close()\n",
    "\n",
    "in_file = open(\"hydrophobicity_values.txt\")\n",
    "vals_lines = in_file.readlines()\n",
    "\n",
    "vals = {}  # Initialise the dictionary\n",
    "for count in range(0, len(vals_lines), 2): \n",
    "\tvals[vals_lines[count].rstrip()] = vals_lines[count+1].rstrip()\t\n",
    "\n",
    "in_file.close()\n",
    "\n",
    "y_data = []\n",
    "for count in range(7, len(seq)-7):\n",
    "\ttotal = 0\n",
    "\tfor count2 in range(-7, 7): \n",
    "\t\ttotal += float(vals[seq[count+count2]])  \n",
    "\taverage = total / 15 \n",
    "\ty_data.append(average)\n",
    "\n",
    "\n",
    "x_data = range(1, len(y_data)+1)\n",
    "\n",
    "plot(x_data, y_data)\n",
    "\n",
    "xlabel(\"Residue number\")\n",
    "ylabel(\"Hydrophobicity\")\n",
    "title(\"Hydrophobicity Plot for \"+seqid)\n",
    "show()\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
